lib/activerecord-bitemporal/scope.rb

```diff
@@ -142,9 +142,10 @@ def bitemporal_value=(value)
   end
 
   module CollectionProxy
-    # Delegate to ActiveRecord::Bitemporal::Relation#bitemporal_value
+    # Delegate to ActiveRecord::Bitemporal::Relation
     # @see https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/associations/collection_proxy.rb#L1115-L1124
-    delegate :bitemporal_value, :bitemporal_value=, to: :scope
+    delegate :bitemporal_value, :bitemporal_value=, :valid_datetime, :valid_date,
+             :transaction_datetime, :bitemporal_option, :bitemporal_option_merge!, to: :scope
```

Comment:
This might be easier to maintain:

```suggestion
    delegate *Relation.public_instance_methods(false), to: :scope
```

It seems that `primary_key`, `build_arel`, and `load` need to be excluded, but to be honest, I don't understand why these need to be excluded from the delegation list.

Reply:
Indeed. I reconsidered that there was no need to exclude them. 

```rb
ActiveRecord::Bitemporal::Relation.public_instance_methods(false)
=>
[:load,
 :bitemporal_option_merge!,
 :valid_datetime,
 :build_arel,
 :primary_key,
 :valid_date,
 :bitemporal_value,
 :transaction_datetime,
 :bitemporal_option,
 :bitemporal_value=]
```

Reply:
Sorry, I was wrong. The `load` needs to be excluded. However, I cannot explain why, so let me think about it for a moment.

Reply:
I checked again and fixed it in d0c2ec8516576307bafc50236fb754e5c53a9273.

This `load` method is `ActiveRecord::Relation#load`.
https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/relation.rb#L740

Inside this, `exec_queries` is called. `exec_query` is `ActiveRecord::Associations::CollectionProxy#exec_queries`.
https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/associations/collection_proxy.rb#L1145

`exec_query` calls `load_target`. Then `load_target` calls `@association`'s `load_target`.
https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/associations/collection_proxy.rb#L44

Calling the `@association`'s method will update the state of `loaded`, etc. Therefore, the `load` method had to be excluded.

The `primary_key` must be delegated to return the correct value (in this case `bitemporal_id`).

The `build_arel` is used to generate queries, but is never used in the case of `CollectionProxy` because it delegates to `scope` or `@association`. However, there was no need to exclude it, so it was delegated.

Reply:
Hmm, since `CollectionProxy` is supposed to be just a proxy, I wonder why it needs to have its own state (e.g. [`CollectionProxy#loaded?` does not refer `@loaded`](https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/associations/collection_proxy.rb#L53-L55)), but I understand that if there is a reason, it should not be delegated.

With that in mind, it might be better to use an allowlist rather than a denylist, since not all `Relation` methods can be delegated (e.g. If we're overriding a method defined for `CollectionProxy`, we shouldn't delegate)

Reply:
> Hmm, since CollectionProxy is supposed to be just a proxy, I wonder why it needs to have its own state (e.g. [CollectionProxy#loaded? does not refer @loaded](https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/associations/collection_proxy.rb#L53-L55))

Yes, `CollectionProxy` is just a proxy, but the `load` methodshould not have been delegated to `scope`. I didn't explain well, but I meant to say that [`load_target`](https://github.com/rails/rails/blob/v7.1.3.4/activerecord/lib/active_record/associations/collection_proxy.rb#L44) and others methods called from the `load` method need to be delegated to `@association` since it refers to the state of `@association`. If the `load` method is delegated to `scope`, the `@association` state will no longer be updated, and as a result `CollectionProxy#loaded?` will not return `true`.

> it might be better to use an allowlist rather than a denylist, since not all Relation methods can be delegated

I too thought it was safer to use an allowlist as I can't imagine what kind of expansion they will do in the future.

Reply:
Fixed.
https://github.com/kufu/activerecord-bitemporal/compare/d0c2ec8..2bb85f7

Reply:
> If the load method is delegated to scope, the `@association` state will no longer be updated

Got it, `CollectionProxy` doesn't need to be stateful, but `Association` do. Delegating to the scope ignores the `association` and as a result state does not propagate properly.